### PR TITLE
[master] add default no-key option to import_matrix, as well as no_header and force_bgz flags.

### PR DESCRIPTION
**Pull request by catoverdrive**
@konradjk some minor tweaks to existing defaults, but otherwise the arguments are the same, plus the no_header flag.

### CODE REVIEW

**Comment by catoverdrive** [comment]
reassigning this because I forgot Jackie was out for the rest of the week

File: src/main/scala/is/hail/io/LoadMatrix.scala [hunk 1]
```diff
@@ -330,15 +353,35 @@ object LoadMatrix {
...
           rv
         }
       }
 
+    val orderedRVD = if (optKeyExpr.isEmpty) {
+      val rvb = new RegionValueBuilder(Region())
+      rvb.start(TArray(TStruct("pk" -> TInt64())))
+      rvb.startArray(partitionCounts.length - 2)
+      var c = 0L
+      partitionCounts.slice(1, partitionCounts.length - 1).foreach { i =>
```
[the remark this answers is not in the document]
Comment (danking) [reply src/main/scala/is/hail/io/LoadMatrix.scala]: I think it would be helpful to pull the bits from `val rvb = ...` to `ranges = new UnsafeIndexedSeq(...)` into a separate method.

This method is already massive as it is. Reading it and understanding all the invariants has been a challenge for me.

[the remark this answers is not in the document]
Comment (danking) [reply src/main/scala/is/hail/io/LoadMatrix.scala]: And ideally we would also not switch on the missingness of the `optKeyExpr` again here, but I don't see a clear path to a cleaner implementation. Worth a think if you have the time. Maybe all the `optKeyExpr` stuff handling stuff needs to get pulled out into one or a few methods. Anything that makes it easier to see, in one place, what happens when there is and is not a key expr.

File: src/test/scala/is/hail/io/ImportMatrixSuite.scala [hunk 2]
```diff
@@ -6,62 +6,118 @@ import is.hail.check.Gen
...
 
+  val genMatrix = VSMSubgen(
+    sSigGen = Gen.const(TString()),
+    saSigGen = Gen.const(TStruct.empty()),
+    vSigGen = Gen.const(TString()),
+    vaSigGen = Gen.const(TStruct.empty()),
+    globalSigGen = Gen.const(TStruct.empty()),
+    tSigGen = Gen.zip(Gen.oneOf[Type](TInt32(), TInt64(), TFloat32(), TFloat64(), TString()), Gen.coin(0.2))
+      .map { case (typ, req) => typ.setRequired(req) }.map { t => TStruct("x" -> t) },
+    sGen = (t: Type) => Gen.identifier.map(s => s: Annotation),
```
Comment (danking) [inline src/test/scala/is/hail/io/ImportMatrixSuite.scala]: This seems dangerous. What happens if the `sSig` isn't a `TString()`?

File: src/test/scala/is/hail/io/ImportMatrixSuite.scala [hunk 3]
```diff
@@ -6,62 +6,118 @@ import is.hail.check.Gen
...
+    saGen = (t: Type) => t.genNonmissingValue,
+    vaGen = (t: Type) => t.genNonmissingValue,
+    globalGen = (t: Type) => t.genNonmissingValue,
+    vGen = (t: Type) => Gen.identifier,
```
Comment (danking) [inline src/test/scala/is/hail/io/ImportMatrixSuite.scala]: also seems dangerous.

File: src/test/scala/is/hail/io/ImportMatrixSuite.scala [hunk 4]
```diff
@@ -6,62 +6,118 @@ import is.hail.check.Gen
...
 import is.hail.expr.types._
+import is.hail.table.Table
 import is.hail.variant.{MatrixTable, VSMSubgen}
 import org.apache.spark.SparkException
 import org.testng.annotations.Test
 import is.hail.utils._
 
 class ImportMatrixSuite extends SparkSuite {
 
+  val genMatrix = VSMSubgen(
```
Comment (danking) [inline src/test/scala/is/hail/io/ImportMatrixSuite.scala]: is there a more descriptive name for this? Is this just the types of matrices that the importer currently supports? Could the rows, cols, row-annos, and col-annos be more generic?

File: src/test/scala/is/hail/io/ImportMatrixSuite.scala [hunk 5]
```diff
@@ -6,62 +6,118 @@ import is.hail.check.Gen
...
-        .map { case (typ, req) => typ.setRequired(req) }.map { t => TStruct("x" -> t) },
-      sGen = (t: Type) => Gen.identifier.map(s => s: Annotation),
-      saGen = (t: Type) => t.genNonmissingValue,
-      vaGen = (t: Type) => t.genNonmissingValue,
-      globalGen = (t: Type) => t.genNonmissingValue,
-      vGen = (t: Type) => Gen.identifier,
-      tGen = (t: Type, v: Annotation) => t.genValue)
-
     forAll(MatrixTable.gen(hc, genMatrix)
       .filter(vsm => !vsm.sampleIds.contains("v"))) { vsm =>
```
Comment (danking) [inline src/test/scala/is/hail/io/ImportMatrixSuite.scala]: this filter shows up in three places, should we pull it up?

File: src/test/scala/is/hail/io/ImportMatrixSuite.scala [hunk 6]
```diff
@@ -6,62 +6,118 @@ import is.hail.check.Gen
...
+      val tmp1 = tmpDir.createTempFile(extension = "vds")
+      vsm.write(tmp1, true)
 
+      val vsm2 = MatrixTable.read(hc, tmp1)
+      assert(vsm.same(vsm2), "not same")
+      true
+    }.check()
+  }
+
+  @Test def testTypesNoKey() {
```
Comment (danking) [inline src/test/scala/is/hail/io/ImportMatrixSuite.scala]: Can we add a test for the case where there are `annotationHeader == None` but we supply a key expression in terms of the `fi`s?

Comment (danking) [reply src/test/scala/is/hail/io/ImportMatrixSuite.scala]: And how about some cases where `noHeader == false` but `annotationHeader == Some(...)`

File: src/test/scala/is/hail/io/ImportMatrixSuite.scala [hunk 7]
```diff
@@ -6,62 +6,118 @@ import is.hail.check.Gen
...
+  }
+
+  @Test def testTypesNoHeader() {
+    forAll(MatrixTable.gen(hc, genMatrix)
+      .filter(vsm => !vsm.sampleIds.contains("v"))) { vsm =>
+
+      val actual2: MatrixTable = {
+        val f = tmpDir.createTempFile(extension = "txt")
+        vsm.makeKT("v = v", "`` = g.x", Array("v")).export(f, header=false)
+        println(f)
```
Comment (danking) [inline src/test/scala/is/hail/io/ImportMatrixSuite.scala]: no debug statements in mainline :)

File: src/test/scala/is/hail/io/ImportMatrixSuite.scala [hunk 8]
```diff
@@ -6,62 +6,118 @@ import is.hail.check.Gen
...
+          .toArray[Annotation])
+          .makeKT("v = v", "`` = g.x", Array())
+          .drop(Array("v"))
+      }
+
+
+      val actual2: MatrixTable = {
+        val f = tmpDir.createTempFile(extension = "txt")
+        getValuesKT(vsm).export(f, header=false)
+        println(f)
```
Comment (danking) [inline src/test/scala/is/hail/io/ImportMatrixSuite.scala]: remove :)

File: src/test/scala/is/hail/io/ImportMatrixSuite.scala [hunk 9]
```diff
@@ -6,62 +6,118 @@ import is.hail.check.Gen
...
+        LoadMatrix(hc, Array(f), None, Seq(), None, cellType = vsm.genotypeSignature, noHeader=true)
+      }
+      println("loaded matrix!")
```
Comment (danking) [inline src/test/scala/is/hail/io/ImportMatrixSuite.scala]: this too

File: src/test/scala/is/hail/io/ImportMatrixSuite.scala [hunk 10]
```diff
@@ -6,62 +6,118 @@ import is.hail.check.Gen
...
+    forAll(MatrixTable.gen(hc, genMatrix)
+      .filter(vsm => !vsm.sampleIds.contains("v"))) { vsm =>
+
+      def getValuesKT(vsm: MatrixTable): Table = {
+        vsm.renameSamples(vsm.sampleIds.zipWithIndex.map { case (id, i) =>
+          "col" + i.toString }
+          .toArray[Annotation])
+          .makeKT("v = v", "`` = g.x", Array())
+          .drop(Array("v"))
+      }
```
Comment (danking) [inline src/test/scala/is/hail/io/ImportMatrixSuite.scala]: lift outside the `forAll` since it doesn't depend on `vsm` directly.

File: src/test/scala/is/hail/io/ImportMatrixSuite.scala [hunk 11]
```diff
@@ -6,62 +6,118 @@ import is.hail.check.Gen
 import is.hail.check.Prop.forAll
```
Comment (danking) [inline src/test/scala/is/hail/io/ImportMatrixSuite.scala]: General comment: I wonder if these tests would look more natural in python? Currently they're a bit wordy and it's hard to parse out exactly what's happening.

File: src/test/scala/is/hail/io/ImportMatrixSuite.scala (hunk 2 above)
Comment (catoverdrive) [reply src/test/scala/is/hail/io/ImportMatrixSuite.scala]: It is, I think? sSigGen and vSigGen are Gen.const(TString()).

File: src/main/scala/is/hail/HailContext.scala [hunk 12]
```diff
@@ -511,25 +511,36 @@ class HailContext private(val sc: SparkContext,
...
+
+    if (forceBGZ)
+      hadoopConf.set("io.compression.codecs",
+        codecs.replaceAllLiterally("org.apache.hadoop.io.compress.GzipCodec", "is.hail.io.compress.BGzipCodecGZ"))
+    try {
+      LoadMatrix(this, inputs, annotationHeaders, annotationTypes, keyExpr, nPartitions = nPartitions,
+        dropSamples = dropSamples, cellType = TStruct("x" -> cellType), missingValue = missingVal, noHeader = noHeader)
+    } finally {
+      hadoopConf.set("io.compression.codecs", codecs)
+    }
```
[the remark this answers is not in the document]
Comment (catoverdrive) [reply src/main/scala/is/hail/HailContext.scala]: For some reason, writing it like this caused the error:

```
hail.utils.java.FatalError: Compression codec is.hail.io.compress.BGZipCodecGZ not found.
```

in both importMatrix and importVCF. I fixed it by not trying to put the `if` statement around the `try` block (and therefore only having the `forceBGZip(force)(body)` method), but I have no idea why this would be the case.

**Comment by catoverdrive** [comment]
@tpoterba @danking fyi I changed the base branch back to master. I don't think it'll be a bad rebase since the changes are pretty much isolated to LoadMatrix.

File: src/test/scala/is/hail/io/ImportMatrixSuite.scala (hunk 2 above)
Comment (catoverdrive) [reply src/test/scala/is/hail/io/ImportMatrixSuite.scala]: (I made vSigGen more generic, and changed vGen--LoadMatrix necessarily imports sampleIDs as TString() right now.)

File: src/test/scala/is/hail/io/ImportMatrixSuite.scala (hunk 6 above)
Comment (catoverdrive) [reply src/test/scala/is/hail/io/ImportMatrixSuite.scala]: `noHeader == false` and `annotationHeader == Some(...)` is kind of difficult because I couldn't find a natural way to export something as a tsv without all the header fields, but I added the other case.

**Comment by catoverdrive** [comment]
@konradjk fyi, I just fixed a bug where missing annotations (but not entries) would be imported as `0` for numeric values and `""` for string values.

File: src/test/scala/is/hail/io/ImportMatrixSuite.scala [hunk 13]
```diff
@@ -1,67 +1,171 @@
...
+
+  def dropRowAnnotations(vsm: MatrixTable, dropFields: Option[Array[String]]=None): MatrixTable =
+    dropFields match {
+      case Some(fields) =>
+        val keep = vsm.vaSignature.fieldNames.filterNot { fn => fields.contains(fn) }
+        val sb = new StringBuilder()
+        keep.foreach { f =>
+          sb ++= s"`$f`: va.`$f`,"
+        }
+        vsm.annotateVariantsExpr(s"va = {${ sb.result().dropRight(1) }}")
```
Comment (danking) [inline src/test/scala/is/hail/io/ImportMatrixSuite.scala]: I think when dealing with operations that happen on master / not-per-element we should go with the simplest to read expression, I think here that's:
```scala
val keepExpr = keep.map(f => s"`$f`: va.`$f`").mkString(",")
vsm.annotateVariantsExpr(s"va = { $keepExpr }")
```

I think in particular the dropRight was a bit odd to me at first (because I was still thinking about Arrays)

---

That said this is also a bit of a nitpick so feel free to ignore.

File: src/test/scala/is/hail/io/ImportMatrixSuite.scala [hunk 14]
```diff
@@ -1,67 +1,171 @@
...
+    kt.export(path, header=header)
+    path
+  }
+
+  def checkValidResult(f: MatrixTable => (MatrixTable, MatrixTable)): Unit = {
+    forAll(MatrixTable.gen(hc, genImportableMatrix)
+      .filter(vsm => vsm.sampleIds.intersect(vsm.vaSignature.fieldNames).isEmpty &&
+        !vsm.vaSignature.fieldNames.contains("v") &&
+        !vsm.sampleIds.contains("v"))) { vsm =>
+      println(vsm.sampleIds.contains("v"))
```
Comment (danking) [inline src/test/scala/is/hail/io/ImportMatrixSuite.scala]: stray println

**Comment by danking** [comment]
@catoverdrive tag me in a comment when you're ready for the final review :)

**Comment by catoverdrive** [comment]
@danking this is ready for a look now!

There's some slight changes to interface, etc to take advantage of the new key design--mostly now it takes row fields instead of an expr. There's also a small fix to some of the partitioner/interval stuff. The python interface is gone now, and I'll put it back in a separate PR, with real docs this time.

And I pulled out all the parsing logic into its own thing.

File: src/main/scala/is/hail/io/LoadMatrix.scala [hunk 15]
```diff
@@ -20,34 +20,23 @@ class LoadMatrixParser(rvb: RegionValueBuilder, fieldTypes: Array[Type], entryTy
...
       if (off > line.length) {
         fatal(
           s"""Error parsing row fields in row $rowNum:
              |    expected $nFields fields but only $ii found.
-             |    in file $file""".stripMargin
+             |    File: $file
+             |    Line:
+             |        ${ line.take(76) }...""".stripMargin
         )
       }
```
Comment (danking) [inline src/main/scala/is/hail/io/LoadMatrix.scala]: I love the ellipsis, but it will be a bit confusing if there's an ellipsis when the string is small. It's kind of hard to find but we have `is.hail.utils.Truncatable`. There are implicit conversions in `is.hail.utils._` that let you call `s.truncate(76)` on any `String` `s`.

File: src/main/scala/is/hail/io/LoadMatrix.scala [hunk 16]
```diff
@@ -57,7 +46,9 @@ class LoadMatrixParser(rvb: RegionValueBuilder, fieldTypes: Array[Type], entryTy
         fatal(
           s"""Incorrect number of entries in row $rowNum:
              |    expected $nCols entries but only $ii entries found.
-             |    in file $file""".stripMargin
+             |    File: $file
+             |    Line:
+             |        ${ line.take(76) }...""".stripMargin
```
Comment (danking) [inline src/main/scala/is/hail/io/LoadMatrix.scala]: same here
